crypto/sm2/sm2_pmeth.c

```diff
@@ -0,0 +1,300 @@
+    EC_GROUP *group;
+
+    switch (type) {
+    case EVP_PKEY_CTRL_EC_PARAMGEN_CURVE_NID:
+        group = EC_GROUP_new_by_curve_name(p1);
+        if (group == NULL) {
+            SM2err(SM2_F_PKEY_SM2_CTRL, SM2_R_INVALID_CURVE);
+            return 0;
+        }
+        EC_GROUP_free(dctx->gen_group);
+        dctx->gen_group = group;
+        return 1;
+
+    case EVP_PKEY_CTRL_EC_PARAM_ENC:
+        if (dctx->gen_group == NULL) {
+            SM2err(SM2_F_PKEY_SM2_CTRL, SM2_R_NO_PARAMETERS_SET);
+            return 0;
+        }
+        EC_GROUP_set_asn1_flag(dctx->gen_group, p1);
+        return 1;
+
+    case EVP_PKEY_CTRL_MD:
+        dctx->md = p2;
+        return 1;
+
+    case EVP_PKEY_CTRL_GET_MD:
+        *(const EVP_MD **)p2 = dctx->md;
+        return 1;
+
+    default:
+        return -2;
+
+    }
+}
+
+static int pkey_sm2_ctrl_str(EVP_PKEY_CTX *ctx,
+                             const char *type, const char *value)
+{
+    if (strcmp(type, "ec_paramgen_curve") == 0) {
+        int nid = NID_undef;
+
+        if (((nid = EC_curve_nist2nid(value)) == NID_undef)
+            && ((nid = OBJ_sn2nid(value)) == NID_undef)
+            && ((nid = OBJ_ln2nid(value)) == NID_undef)) {
+            SM2err(SM2_F_PKEY_SM2_CTRL_STR, SM2_R_INVALID_CURVE);
+            return 0;
+        }
+        return EVP_PKEY_CTX_set_ec_paramgen_curve_nid(ctx, nid);
+    } else if (strcmp(type, "ec_param_enc") == 0) {
+        int param_enc;
+
+        if (strcmp(value, "explicit") == 0)
+            param_enc = 0;
+        else if (strcmp(value, "named_curve") == 0)
+            param_enc = OPENSSL_EC_NAMED_CURVE;
+        else
+            return -2;
+        return EVP_PKEY_CTX_set_ec_param_enc(ctx, param_enc);
+    }
+
+    return -2;
+}
+
+static int pkey_sm2_paramgen(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey)
+{
+    EC_KEY *ec = NULL;
+    SM2_PKEY_CTX *dctx = ctx->data;
+    int ret = 0;
+
+    if (dctx->gen_group == NULL) {
+        SM2err(SM2_F_PKEY_SM2_PARAMGEN, SM2_R_NO_PARAMETERS_SET);
+        return 0;
+    }
+    ec = EC_KEY_new();
+    if (ec == NULL)
+        return 0;
+    ret = EC_KEY_set_group(ec, dctx->gen_group);
+    if (ret)
+        EVP_PKEY_assign_EC_KEY(pkey, ec);
+    else
+        EC_KEY_free(ec);
+    return ret;
+}
+
+static int pkey_sm2_keygen(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey)
+{
+    EC_KEY *ec = NULL;
+    SM2_PKEY_CTX *dctx = ctx->data;
+    const EC_GROUP *group = NULL;
+
+    if (dctx->gen_group) {
+        group = dctx->gen_group;
+    }
+    else if (ctx->pkey) {
+        group = EC_KEY_get0_group(ctx->pkey->pkey.ec);
+    }
+    else {
+        SM2err(SM2_F_PKEY_SM2_KEYGEN, SM2_R_NO_PARAMETERS_SET);
+        return 0;
+    }
+
+    ec = EC_KEY_new();
+    if (ec == NULL)
+        return 0;
+
+    if (!EC_KEY_set_group(ec, group)) {
+        EC_KEY_free(ec);
+        return 0;
+    }
+
+    if(!EC_KEY_generate_key(ec)) {
+        EC_KEY_free(ec);
+        return 0;
+    }
+
+    return EVP_PKEY_assign_EC_KEY(pkey, ec);
+}
```

Comment:
I'd like to suggest "ultimate" solution, **omit** pkey_sm2_paramgen and pkey_sm2_keygen. There is no any SM2 "twist" to keygen and default one from common pmeth would work [with the curve in question]. Application would even have to do anything special. It should also be noted that these two are not exercised in any way, so we don't actually know if it works...

Reply:
> Application would even have to do anything special.

I meant to write "application would **not** even have to do anything special."

Reply:
Just to clarify, there actually are more problems with paramgen and keygen. But instead of addressing the problems suggestion is to omit the functions in question from sm2_pmeth. This doesn't affect user's ability to generate keys, because it can be done with "standard" ec_pmeth. If at later point it turns out that sm2_pmeth has to have these two, then they can be added then. Preferably as "bypass" to ec_pmeth. By "bypass" I mean that since sm2_pmeth is an "alias" pmeth, it would be appropriate to engage "parent" pmeth for functions that are not performing method-specific operations.

Just in case. The fact that there are more problems with paramgen and keygen actually means that there are problems even in ec_pmeth. I plan to address them as soon as we are ready with this one.